### PR TITLE
add possibility to download to OCI image-layout

### PR DESCRIPTION
**Pull request by runcom**
replacement for #102 - currently blocked on containers/image#23 bits (this branch contains code from containers/image#23 from my personal fork) 

/cc @mtrmac 

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

### CODE REVIEW

**Comment by runcom** [comment]
there something wrong with this I cannot spot

**Comment by runcom** [comment]
so here is the funny thing - sending v2s1 in accept headers isn't returning anymore a v2s1 but always a v2s2...

**Comment by runcom** [comment]
alright, having a default mime types list in `FromSource` is breaking this PR as opposed to #102 which was working because the case 1) docker -> dir pulls a v2s2 manifest (because dir don't care) and 2) dir -> atomic breaks because it wants v2s1

**Comment by runcom** [comment]
this was the relevant conversation https://github.com/containers/image/pull/26#discussion_r68958299

**Comment by mtrmac** [comment]
Ouch, that is painful. OTOH the ability to get v2s2, or anything else, into a `dir:` is valuable, and “copying to a `dir:` pulls v2s1 because otherwise our test would break” is not a good rationale for our users :)

For now, I guess let’s run the streaming test with a known-v2s1 image?

Longer-term we might want to allow the user to choose what to use in `copy` (this applies primarily to manifest lists and the arch dependencies, but that infrastructure or something very similar could be used to force using specific formats)—but that is clearly not something we need or want to build right now.

File: cmd/skopeo/layers.go [hunk 1]
```diff
@@ -18,7 +19,12 @@ var layersCmd = cli.Command{
 		if err != nil {
 			return err
 		}
-		src := image.FromSource(rawSource)
+		src := image.FromSource(rawSource, []string{
+			// TODO: skopeo layers only support these now
+			// eventually we'll remove this command altogether...
+			manifest.DockerV2Schema1SignedMIMEType,
+			manifest.DockerV2Schema1MIMEType,
+		})
```
Comment (mtrmac) [inline cmd/skopeo/layers.go]: Really? AFAICS the code is generic and works for everything.

**Comment by mtrmac** [comment]
Also it Would Be Nice™ to have an integration test for OCI, but that is absolutely not a blocker for me.

File: cmd/skopeo/layers.go (hunk 1 above)
Comment (runcom) [inline cmd/skopeo/layers.go]: I think the goal is to deprecate this command, or blindly call out copy

Comment (mtrmac) [inline cmd/skopeo/layers.go]: Yeah, sure, but that is not a rationale for this change to `layers.go` nor does it explain how “skopeo layers only support these now”.

Or are you saying that the code would work fine but you want to intentionally restrict it to v2s1 to force users to use `skopeo copy`?

Comment (runcom) [inline cmd/skopeo/layers.go]: The former :) atomic uses layers command with v2s1 right now, so I won't change anything let's make sure to fix copy instead also

**Comment by runcom** [comment]
I will add it as soon as we Marge the changes in containers/image

**Comment by runcom** [comment]
The issue however is real because official docker images are getting rebuilt as we speak and they all will be v2s2 and docker-dir dir-atomic will always be broken (while I think Docker-atomic isn't broken)

**Comment by mtrmac** [comment]
I don’t really care about the `docker`→`dir`→`atomic` workflow as an use case; as far as I am concerned, `dir` is primarily a debugging tool, or at most an internal implementation detail of `atomic pull`.  (Is there consensus on this?)

If `skopeo copy` does the right thing for `docker`→`atomic`, that’s more important.

(Also note that the test doesn’t test `docker`→`dir`→`atomic`, it just assumes that `docker`→ will always return the same image, which is no longer the case.  (Technically it was never the case, that test breaks if the content on the Docker Hub changes between the two commands, but I digress…) The purpose of the test is to show that `skopeo copy` does not modify the image; it is perfectly fine to demonstrate that with v2s1, though we could eventually in some vague future have such a test for every MIME type.)

Also, over time OpenShift will, hopefully, add v2s2 support, and most of the issue will disappear or at least become less noticeable.

**Comment by runcom** [comment]
Awesome I'm less worried now thanks

**Comment by mtrmac** [comment]
I think you found some v2s1 images (or perhaps there are some intentionally v2s1 images?) for the other test experiments; if that image is (and will remain?) v2s1, the streaming test with that image should succeed IIUC.

**Comment by runcom** [comment]
It will, I'm about to fix this in a moment

**Comment by runcom** [comment]
@mtrmac PTAL - I've added a small smoke test for Docker->OCI - I'll re-iterate on tests later in followup PRs since I'm taking care of others as well

This should be good to merge after tests

**Comment by runcom** [comment]
wait, this needs containers/image#28

File: integration/copy_test.go [hunk 2]
```diff
@@ -55,14 +55,23 @@ func (s *CopySuite) TestCopySimple(c *check.C) {
...
 	assertSkopeoSucceeds(c, "", "--debug", "copy", "dir:"+dir1, "atomic:myns/unsigned:unsigned")
 	// The result of pushing and pulling is an unmodified image.
 	assertSkopeoSucceeds(c, "", "copy", "atomic:myns/unsigned:unsigned", "dir:"+dir2)
 	out := combinedOutputOfCommand(c, "diff", "-urN", dir1, dir2)
 	c.Assert(out, check.Equals, "")
 
+	// docker v2s2 -> OCI image layout
+	// ociDest will be created by oci: if it doesn't exist
+	// so don't create it here to exercise auto-creation
+	ociDest := "busybox-lateset"
```
Comment (mtrmac) [inline integration/copy_test.go]: s/lateset/latest/

**Comment by mtrmac** [comment]
Great, thanks for making that work!; LGTM after re-vendoring from containers/image master.

**Comment by runcom** [comment]
for some reason Travis failed integration tests on `TestCopy*` but running test locally everything is fine.

**Comment by runcom** [comment]
@mtrmac 
```
time="2016-07-01T07:42:58Z" level=debug msg="Error body {\"errors\":[{\"code\":\"UNKNOWN\",\"message\":\"unknown error\",\"detail\":{\"ErrStatus\":{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"ImageStreamMapping \\\"unsigned\\\" is invalid: tag: Required value\",\"reason\":\"Invalid\",\"details\":{\"name\":\"unsigned\",\"kind\":\"ImageStreamMapping\",\"causes\":[{\"reason\":\"FieldValueRequired\",\"message\":\"Required value\",\"field\":\"tag\"}]},\"code\":422}}}]}\n" 
```
do you know what can cause this in Travis while locally is fine?

**Comment by runcom** [comment]
FWIW I'm not lying locally it works:
```
+ go test -test.timeout=10m -check.v github.com/projectatomic/skopeo/integration
SKIP: check_test.go:82: SkopeoSuite.TestCanAuthToPrivateRegistryV2WithoutDockerCfg (we need to restore --username --password flags!)
SKIP: check_test.go:89: SkopeoSuite.TestNeedAuthToPrivateRegistryV2WithoutDockerCfg (mock empty docker config)
PASS: check_test.go:98: SkopeoSuite.TestNoNeedAuthToPrivateRegistryV2ImageNotFound	0.011s
PASS: check_test.go:73: SkopeoSuite.TestVersion	0.005s
PASS: copy_test.go:48: CopySuite.TestCopySimple	11.175s
PASS: copy_test.go:79: CopySuite.TestCopyStreaming	10.946s
PASS: signing_test.go:64: SigningSuite.TestSignVerifySmoke	2.825s
OK: 5 passed, 2 skipped
PASS
ok  	github.com/projectatomic/skopeo/integration	36.931s
```

**Comment by mtrmac** [comment]
Some work is happening on the OpenShift PR, changing the API; so probably the server has changed and broken the tests.

You very likely still have a locally built (or locally cached) container with the old OpenShift version, so tests work for you, but Travis rebuilds with no cache and uses the new version.

I was promised the current API would be archived in another branch, so we can just switch until we update the client, let me have a look…

**Comment by mtrmac** [comment]
> I was promised the current API would be archived in another branch, so we can just switch until we update the client, let me have a look…

I can’t find that archive, so I rescued the old commit from one of my Docker images: #139 .

**Comment by runcom** [comment]
all good! merging finally :) Thanks for the help @mtrmac

**Pull request merged**